modules/openxr/extensions/openxr_opengl_extension.cpp

```diff
@@ -0,0 +1,479 @@
+		print_line("OpenXR: Requested OpenGL version exceeds the maximum version this runtime has been tested on and is known to support.");
+		print_line("- desired_version ", OpenXRUtil::make_xr_version_string(p_desired_version));
+		print_line("- minApiVersionSupported ", OpenXRUtil::make_xr_version_string(opengl_requirements.minApiVersionSupported));
+		print_line("- maxApiVersionSupported ", OpenXRUtil::make_xr_version_string(opengl_requirements.maxApiVersionSupported));
+	}
+
+	return true;
+}
+
+#ifdef WIN32
+XrGraphicsBindingOpenGLWin32KHR OpenXROpenGLExtension::graphics_binding_gl;
+#elif ANDROID_ENABLED
+XrGraphicsBindingOpenGLESAndroidKHR OpenXROpenGLExtension::graphics_binding_gl;
+#else
+XrGraphicsBindingOpenGLXlibKHR OpenXROpenGLExtension::graphics_binding_gl;
+#endif
+
+void *OpenXROpenGLExtension::set_session_create_and_get_next_pointer(void *p_next_pointer) {
+	XrVersion desired_version = XR_MAKE_VERSION(3, 3, 0);
+
+	if (!check_graphics_api_support(desired_version)) {
+		print_line("OpenXR: Trying to initialize with OpenGL anyway...");
+		//return p_next_pointer;
+	}
+
+	DisplayServer *display_server = DisplayServer::get_singleton();
+
+#ifdef WIN32
+	graphics_binding_gl.type = XR_TYPE_GRAPHICS_BINDING_OPENGL_WIN32_KHR,
+	graphics_binding_gl.next = p_next_pointer;
+
+	graphics_binding_gl.hDC = (HDC)display_server->window_get_native_handle(DisplayServer::WINDOW_VIEW);
+	graphics_binding_gl.hGLRC = (HGLRC)display_server->window_get_native_handle(DisplayServer::OPENGL_CONTEXT);
+#elif ANDROID_ENABLED
+	graphics_binding_gl.type = XR_TYPE_GRAPHICS_BINDING_OPENGL_ES_ANDROID_KHR;
+	graphics_binding_gl.next = p_next_pointer;
+
+	graphics_binding_gl.display = eglGetCurrentDisplay();
+	graphics_binding_gl.config = (EGLConfig)0; // https://github.com/KhronosGroup/OpenXR-SDK-Source/blob/master/src/tests/hello_xr/graphicsplugin_opengles.cpp#L122
+	graphics_binding_gl.context = eglGetCurrentContext();
+#else
+	graphics_binding_gl.type = XR_TYPE_GRAPHICS_BINDING_OPENGL_XLIB_KHR;
+	graphics_binding_gl.next = p_next_pointer;
+
+	void *display_handle = (void *)display_server->window_get_native_handle(DisplayServer::DISPLAY_HANDLE);
+	void *glxcontext_handle = (void *)display_server->window_get_native_handle(DisplayServer::OPENGL_CONTEXT);
+	void *glxdrawable_handle = (void *)display_server->window_get_native_handle(DisplayServer::WINDOW_HANDLE);
+
+	graphics_binding_gl.xDisplay = (Display *)display_handle;
+	graphics_binding_gl.glxContext = (GLXContext)glxcontext_handle;
+	graphics_binding_gl.glxDrawable = (GLXDrawable)glxdrawable_handle;
+
+	if (graphics_binding_gl.xDisplay == nullptr) {
+		print_line("OpenXR Failed to get xDisplay from Godot, using XOpenDisplay(nullptr)");
+		graphics_binding_gl.xDisplay = XOpenDisplay(nullptr);
+	}
+	if (graphics_binding_gl.glxContext == nullptr) {
+		print_line("OpenXR Failed to get glxContext from Godot, using glXGetCurrentContext()");
+		graphics_binding_gl.glxContext = glXGetCurrentContext();
```

Comment:
That's a good question! I copied this code from the OpenXR plugin for Godot 3. I think there may have been a time before `OS::get_native_handle()` (the name of the function in Godot 3) was added to Godot, and this code was a fallback for if you used the plugin with an older version of Godot. We should probably just remove all those fallbacks, since we know we can rely on `DisplayServer::window_get_native_handle()`. And now that in Godot 4 we have multiple window support, it's more likely that calling `glXGetCurrentContext()` would give us the incorrect context anyway

Reply:
Yeah, I found the "paper trail" for this!

PR #42531 added `OS::get_native_handle()` added to Godot 3, so that https://github.com/GodotVR/godot_openxr/pull/6 could use it in the OpenXR plugin.

I'll make a PR to remove the redundant fallback code...

Reply:
And here's the PR to remove it: https://github.com/godotengine/godot/pull/68453